lib/datadog/tracing/client_ip.rb

```diff
@@ -0,0 +1,191 @@
+# typed: true
+
+require 'ipaddr'
+
+require_relative '../core/configuration'
+require_relative 'metadata/ext'
+require_relative 'span'
+
+module Datadog
+  module Tracing
+    # Common functions for supporting the `http.client_ip` span attribute.
+    module ClientIp
+      DEFAULT_IP_HEADERS_NAMES = %w[
+        x-forwarded-for
+        x-real-ip
+        x-client-ip
+        x-forwarded
+        x-cluster-client-ip
+        forwarded-for
+        forwarded
+        via
+        true-client-ip
+      ].freeze
+
+      # A collection of headers.
+      class HeaderCollection
+        # Gets a single value of the header with the given name, case insensitive.
+        #
+        # @param [String] header_name Name of the header to get the value of.
+        # @returns [String, nil] A single value of the header, or nil if the header with
+        #   the given name is missing from the collection.
+        def get(header_name)
+          nil
+        end
+
+        def self.from_hash(hash)
+          HashHeaderCollection.new(hash)
+        end
+      end
+
+      # Sets the `http.client_ip` tag on the given span.
+      #
+      # This function respects the user's settings: if they disable the client IP tagging,
+      # or provide a different IP header name.
+      #
+      # @param [Span] span The span that's associated with the request.
+      # @param [HeaderCollection, #get, nil] headers A collection with the request headers.
+      # @param [String, nil] remote_ip The remote IP the request associated with the span is sent to.
+      def self.set_client_ip_tag(span, headers, remote_ip)
+        return if configuration.disabled
+
+        ip = client_address_from_request(headers, remote_ip)
+        if !configuration.header_name && ip.nil?
+          header_names = ip_headers(headers).keys
+          span.set_tag(TAG_MULTIPLE_IP_HEADERS, header_names.join(',')) unless header_names.empty?
+          return
+        end
+
+        unless valid_ip?(ip)
```

Comment:
I feel this nesting of `unless` is confusing.

Is this necessary since `extract_ip_from_full_address` seems to return the address when its internal criteria don't match?

Reply:
I completely refactored that part, let me know if it makes more sense now :)